src/madsci_client/madsci/client/data_client.py

```diff
@@ -103,11 +212,92 @@ def get_datapoints(self, number: int = 10) -> list[DataPoint]:
         response.raise_for_status()
         return [DataPoint.discriminate(datapoint) for datapoint in response.json()]
 
+    def query_datapoints(self, selector: Any) -> dict[str, DataPoint]:
+        """Query datapoints based on a selector."""
+        if self.url is None:
+            return {
+                datapoint_id: datapoint
+                for datapoint_id, datapoint in self._local_datapoints.items()
+                if selector(datapoint)
+            }
+        response = requests.post(
+            f"{self.url}datapoints/query", json=selector, timeout=10
+        )
+        response.raise_for_status()
+        return {
+            datapoint_id: DataPoint.discriminate(datapoint)
+            for datapoint_id, datapoint in response.json().items()
+        }
+
     def submit_datapoint(self, datapoint: DataPoint) -> DataPoint:
-        """Submit a Datapoint object"""
+        """Submit a Datapoint object.
+
+        If object storage is configured and the datapoint is a file type,
+        the file will be automatically uploaded to object storage instead
+        of being sent to the Data Manager server.
+
+        Args:
+            datapoint: The datapoint to submit
+
+        Returns:
+            The submitted datapoint with server-assigned IDs if applicable
+        """
+        # Case 1: Handle ObjectStorageDataPoint with path directly
+        if (
+            hasattr(datapoint, "data_type")
```

Comment:
You shouldn't have to do this hasattr checks, given all DataPoint objects are required to have a data_type field